rfcs/unboxed-types.md

```diff
@@ -0,0 +1,458 @@
+not `string` and `#float`.
+
+## Unboxed tuples and records
+
+The second part of this proposal is to allow user-defined unboxed
+types (by contrast to what's above: user-defined boxed types
+containing unboxed fields). The simplest of these is unboxed tuples:
+
+    #( int * float * string )
+
+This is an unboxed tuple, and is passed around as three separate
+values.  Passing it to a function requires three registers.
+The layout of an unboxed tuple is the product of the layout of its
+components. In particular, the layout of `#( string * string )` is
+`value * value`, not `value`, and so it cannot be stored in, say, a
+list.
+
+Unboxed records are in theory just as straightforward, but have some
+extra complexity just to handle the various ways that types are
+defined and used. Suppose you write:
+
+    type t = { x : int32; y : int32 }
+
+Sometimes we want to refer to `t` as a boxed type, to store it in
+lists and other standard data structures. Sometimes we want to refer
+to `t` as an unboxed type, to avoid extra allocations (in particular,
+to inline it into other records). The language never infers which is
+which (boxing and allocation is always explicit in this design), so we
+need different syntax for both.  So, `type t = { ... }` defines *two*
+types, the boxed record `t : value` and the unboxed record `#t :
+int32 * int32`. They have different syntax for introduction and
+elimination:
+
+    let p : t = { x = a; y = b }
+    let px = p.x
+    let q : #t = #{ x = a; y = b }
+    let qx = q.#x
+
+An unboxed record like `q` cannot itself contain mutable
+fields. However, a boxed record can contain a mutable field of type
+`q`. This can be used to inline one record inside another, e.g.:
+
+    type s = { foo : int; mutable bar : #t }
+
+This defines a boxed record, layed out as three fields. We will
+support efficient updates of sub-parts of `bar`, at least with the
+following syntax:
+
+    x.bar <- { x.bar with y = y' }
+
+It would be nice to also allow the syntax `x.bar.#y <- y'`, although
+this is a little trickier to understand and implement, since it is
+`x.bar` rather than `x.bar.y` which is the mutable field.
+
+Type aliases do not automatically bind an unboxed variant, because
+the scoping of type aliases should not depend on whether they expand
+to records or not. So, if you write:
+
+    type u = t
+
+then you don't get `#u`. You can get around this by re-exporting the
+definition:
+
+    type u = t = { x : int32; y : int32 }
+
+which gets you both `u = t` and `#u = #t`. You can also define an
+alias directly to the unboxed type:
+
+    type u = #t
+
+Similarly, an abstract type abstracts only the boxed type:
+
+    module M : sig
+      type t
+    end = struct
+      type t = { x : int32; y : int32 }
+    end
+
+defines `M.t` but not `M.#t`. (It may be useful to allow exporting
+`M.#t` despite `t` being abstract, perhaps with syntax like `type t :
+value with #t : int32 * int32`).
```

Comment:
@gasche says
If this exposes the fact that box(#t) = t, then is t still abstract in any reasonable sense? This seems a bit fragile.

 @stedolan says
Right, I'm not sure how useful this is. The motivation is this: perhaps I'd like to expose a type in both boxed and unboxed form (t and #t), so that a client of the library can freely box and unbox the type (perhaps storing it inlined inside another data structure), but cannot otherwise introduce or eliminate t except by using the library's interface.

To let the client box and unbox t, the client must know that #t : int32 * int32. This lets the client compile code that passes around #t (knowing that it takes two integer registers, 8 bytes of memory, and should not be registered with the GC), but the client cannot do anything else with it.

@gasche says
That makes perfect sense -- you already explain how having incompatible types of layout int32, this is just an abstract type at layout int32. Thanks for the clarification.

@lpw25 says
Note that in the current proposal, type t with #t : l does not actually expose that box(#t) = t in any meaningful sense because there is no box type constructor and no builtin operations that allow you to make a t from a #t. The syntax merely allows the type #t to be exposed abstractly in the interface.

@gasche says
Wouldn't be useful to think of immediate values as boxed in the same sense (so type t [@@immediate] actually means type t : value with #t : immediate rather than type t : immediate)? Can we get rid of subtyping that way?

More precisely, I think that #t in that context would naturally represent the untagged form of the immediate value, rather than just the immediate value itself.

@stedolan says
Reusing this syntax to allow explicit tagging/untagging for immediate types is a nice idea. I don't see how this replaces the immediate <= value subtyping, though: that's not about adding and removing tagging but about the inclusion of tagged integers into values.

@gasche says
The idea would be to never ask "are the value of this type immediate values?", but instead "is this type known to be the tagging of an untagged type?", when making optimization decisions. (Similarly to the fact that exporting type t with #t : int32 should let consumers deduce that the values of t are pointeres.) If that is enough to make the same optimization decisions, then we can drop the current concept of "immediate type" -- which could remove the need for having layout subtyping.

@stedolan says
Right, that makes sense now. The idea is that rather than allowing kindings t : int, you can instead express t : value = tag(#t) and #t : untagged_int. I think this would work, but it seems more awkward than the subkinding approach for a couple of reasons:

It's a bit more awkward to quantify over immediate types this way. This is useful when writing records with mutable fields that guarantee no uses of caml_modify:

```type ('a : immediate) r = { mutable x : 'a }```

becomes

```type ('a : untagged_int) r = { mutable x : tag('a) }```

It's no longer possible to instantiate this definition as int r: instead, the user must know that the builtin int type is defined as tag(#int) and write #int r (even though there are no uses of #int in the program). Similarly, the user must know about the untagged versions of variant types that have only constant constructors.

This trick stops working once we have more than one property of interest. For instance, if we implement efficient nullable types (as discussed below), then we also want to track whether a type may include a null value. We cannot define int as both tag(#int) and not_null(#int') unless we do some quite hairy type system work to commute tag and not_null.

I was hoping to make all compilation decisions about how a type should be stored (how many registers it takes, whether it needs to be registered with the GC, etc.) by examining the type's layout. However, this approach would mean a mixture of examining the type's layout (to see if it's value) and then matching on the type itself (to see if it's tag(#foo)) in order to make these decisions.

It sounds like the motivation for this instead of an immediate kind is the perceived "fishiness" of the algorithm for unification with layouts and subkinding. My informal description of the algorithm below is full of mutation, because that's how unification is implemented in OCaml. I'll write up a syntactic presentation of the unification algorithm, and see if you find that less fishy.

@gasche says
What about
```type 'a r = { mutable x : 'a } constraint 'a = tag('i)```
? Just kidding. I think your second point is interesting and the third is convincing.

I was critical of the algorithm description (and I do think it could be improved), but that was not necessarily the point for my consideration above; I just think that subtyping tend to introduce complexity, and that it's interesting to seriously check that we want it in the design. (I realize you are an interesting choice of person to complain to about subtyping.)

In the big picture I'm fairly confident that you can propose a good design even in presence of subtyping, because the problem domain is constrained: we don't necessarily need to do a lot of inference, and for now there is no interaction between subtyping and polymorphism -- given that there is no kind polymorphism, for now.

@stedolan says
Hah! I didn't think of using constraints here, nice.

Here's a syntactic version of the unification algorithm with subkinding, hopefully it seems less fishy than before: https://gist.github.com/stedolan/349b146b4744549fdb25088c4501de73. (Just the basic algorithm for now, I'll write up the details of handling type aliases later. Again, the story is the same as for standard unification: in principle, you can expand aliases beforehand, but in practice you're better off expanding them only when needed and caching the results)

@gasche says
Thanks! The description is very clear.

In the Instr rule, you check Ahat : kappa, using the kind of the alpha variable. Don't you rather want to lower the kind of alpha if Ahat can be given a smaller kind? That's what I understood from the informal description.

@gasche says
(And: I agree that having the rules for parametrized type aliases would be nice.)

@gasche says
Thinking aloud more: if you implement kind lowering during unification based on kind checking (and not just the intersection of variable's declared kinds as you do now), you probably want to have a more algorithmic presentation of kind checking Gamma |- A => kappa, which computes a kind that is minimal with respect to the declarative check Gamma |- A : kappa. (On a finite kind domain, we can always compute this by taking the lub of all admissible kinds, but this is less informative than an algorithmic judgment that is close to the implementation.)

@stedolan says
In the Instr rule, you check Ahat : kappa, using the kind of the alpha variable. Don't you rather want to lower the kind of alpha if Ahat can be given a smaller kind?

OK, there's some misunderstanding here. (Showing that It was useful to poke me to write a more precise description!) We don't lower the kind of alpha if Ahat can be given a smaller kind, because alpha is removed (substituted away). Only uninstantiated variables have kinds, because only uninstantiated variables range over anything. Once a variable is instantiated, we have neither a variable nor a kind any more.

When I talk about mutating the kind in the informal description, I'm describing the rule Eq, where variables alpha and beta are removed and replaced with a new variable gamma, ranging over the intersection of the layouts. (In an actual implementation, it is convenient to choose the name gamma = alpha, so all you need to do is mutate alpha's kind and delete beta).

But you're right that the rules should specify the kinds of the variables that are uninstantiated. I've currently written the output of unification as a list of variable assignments, but it should be a pair of a list of variable assignments and a list of unassigned variables (with kinds for the latter). That is, for each variable in Gamma to begin with, the output either gives a definition of that variable (if it was instantiated), or a new kind for that variable (if it was not).